### PR TITLE
first draft of JWT token auth update

### PR DESCRIPTION
**Pull request by mystiberry-okta**
## Description:
- First draft of doc for JWT token authentication

## Type of Pull Request:

- [x] Content (Documentation updates or typo-fixes)
- [ ] Blog Post
- [ ] Functional (CSS changes, JS updates, or layout modifications)

### Resolves:
* [OKTA-131109](https://oktainc.atlassian.net/browse/OKTA-131109)

## How Has This Been Tested?
- [x] I have built this locally and verified that it does not break existing functionality.
- [ ] For functional changes, I have tested against Chrome, Firefox, and Safari, and mobile.
- [ ] For functional changes, I have added tests.

## Primary Reviewers:
@larsjohansen-okta There are gaps, not ready for the whole team or an editor to review. I'm sure there are errors.

@larsjohansen-okta not sure I captured the client_assertion/client_assertion_type correctly, or in all the right places. Not sure if we need more content around JWT/HMAC etc.

### CODE REVIEW

File: _source/_docs/api/resources/oauth2.md [hunk 1]
```diff
@@ -194,18 +194,20 @@ This API returns Access Tokens, ID Tokens, and Refresh Tokens, depending on the
...






[PKCE](#request-parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.   | String |
+| client_id        | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret` to authenticate the client application.                                                                                                                                              | String |
+| client_secret    | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.                                                                                                                                              | String |
+| client_assertion | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: I think we should have a link to more info about Token Authentication Methods here, as it's not obvious how the JWT should be constructed.

File: _source/_docs/api/resources/oauth2.md [hunk 2]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+  in an Authorization header in the Basic auth scheme (`client_secret_basic`). For authentication with Basic auth, an HTTP header with the following format must be provided with the POST request:
+  ~~~sh
+  Authorization: Basic ${Base64(<client_id>:<client_secret>)}
+  ~~~
+* Provide [`client_id`](oidc.html#request-parameters) and [`client_secret`](https://support.okta.com/help/articles/Knowledge_Article/Using-OpenID-Connect)
+  as additional parameters to the POST body (`client_secret_post`)
+* Provide [`client_id`](oidc.html#request-parameters) in a JWT that you sign with the [`client_secret`](https://support.okta.com/help/articles/Knowledge_Article/Using-OpenID-Connect)
+  using HMAC algorithms HS256, HS384, or HS512. Specify the JWT in `client_assertion` and the type, `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`, in `client_assertion_type` in the request. 
+
+Including these credentials in both the Authorization header and POST body is not allowed.
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: This is slightly confusing; the point of this sentence should be to show that it is not possible to provide more than just 1 out of the 3 ways to authenticate the client, Authorization header is explained using 1 way and POST body is technically both of the latter approaches as they are both in the POST body, perhaps something along these lines:

It is only allowed to include these credentials using one of the methods described.

File: _source/_docs/api/resources/oauth2.md [hunk 3]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+
+>If the value returned by `token_endpoint_authentication` in the [OAuth 2.0 Clients API](/docs/api/resources/oauth-clients.html#update-client-application) is not what you wish to use, you can change the value of a client app's `token_endpoint_authentication` with any of the values returned by `token_endpoint_auth_methods_support` (`client_secret_post`, `client_secret_basic`, or `client_secret_jwt`).
+You can't change this value in the Okta user interface.
+
+##### Request Parameters for JWT Token Authentication
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: ##### Request Parameters for JWT Token Authentication -> ##### Request Parameters for Client Secret JWT Client Authentication

File: _source/_docs/api/resources/oauth2.md [hunk 4]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+
+If you use a JWT for token authentication (`client_secret_jwt`), use the following request parameters:
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "JWT" -> "Client Secret JWT"

Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "token authentication" -> "client authentication"

File: _source/_docs/api/resources/oauth2.md [hunk 5]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+
+| Parameter | Description                                                                      | Type   |
+|:----------|:---------------------------------------------------------------------------------|:-------|
+| exp       | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat       | Required. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: This claim is optional, but will be validated if provided.

File: _source/_docs/api/resources/oauth2.md [hunk 6]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+| sub       | Required. The subject of the token.                                              | String |
+| aud       | Required. The full URL of the resource you're using the JWT to access.           | String |
+| iss       | Required. The issuer of the token.                                               | String |
+| jti       | Optional. The identifier of the token.                                           | String |
+
+Parameter Details
+
+* If `jti` is specified, the token can only be used once. So, for example, subsequent refreshes won't succeed.
+* If `exp` is specified, be aware that requests will fail if the time specified is more than an hour before the request is received.
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: This is required, so we should assume that it is specified.

File: _source/_docs/api/resources/oauth2.md [hunk 7]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+| exp       | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat       | Required. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
+| sub       | Required. The subject of the token.                                              | String |
+| aud       | Required. The full URL of the resource you're using the JWT to access.           | String |
+| iss       | Required. The issuer of the token.                                               | String |
+| jti       | Optional. The identifier of the token.                                           | String |
+
+Parameter Details
+
+* If `jti` is specified, the token can only be used once. So, for example, subsequent refreshes won't succeed.
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "subsequent refreshes won't succeed." -> "subsequent token requests won't succeed."

File: _source/_docs/api/resources/oauth2.md [hunk 8]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+* If `exp` is specified, be aware that requests will fail if the time specified is more than an hour before the request is received.
+* If `iat` is specified, must be a time before the request is received.
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "If `iat` is specified, must be" -> "If `iat` is specified, then it must be"

File: _source/_docs/api/resources/oauth2.md (hunk 6 above)
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "The `exp` claim will fail the request if the expiration time is more than 1 hour into the future or has expired."

File: _source/_docs/api/resources/oauth2.md [hunk 9]
```diff
@@ -359,6 +390,8 @@ The following parameters can be posted as a part of the URL-encoded form values
 | token_type_hint | A hint of the type of `token`.                                                                                                                                                                 | String |
 | client_id       | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret`  to authenticate the client application. | String |
 | client_secret   | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.  | String |
+| client_assertion | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: Same comment here, link to the detailed section.

File: _source/_docs/api/resources/oauth2.md [hunk 10]
```diff
@@ -476,38 +499,30 @@ The API takes an Access Token or Refresh Token and revokes it. Revoked tokens ar
...
+|:-----------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------|
+| token            | An access token or refresh token.                                                                                                                                | String |
+| token_type_hint  | A hint of the type of *token*.                                                                                                                                   | String |
+| client_id        | The client ID generated as a part of client registration. This is used in conjunction with the *client_secret* parameter to authenticate the client application. | String |
+| client_secret    | The client secret generated as a part of client registration. This is used in conjunction with the *client_id* parameter to authenticate the client application. | String |
+| client_assertion | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: Same comment here, link to the detailed section.

File: _source/_docs/api/resources/oidc.md [hunk 11]
```diff
@@ -356,24 +356,51 @@ An implicit client can only introspect its own tokens, while a confidential clie
...
+| token                 | An Access Token, ID Token, or Refresh Token.                                                                                                                                                   | String |
+| token_type_hint       | A hint of the type of *token*.                                                                                                                                                                 | String |
+| client_id             | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret`  to authenticate the client application. | String |
+| client_secret         | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.  | String |
+| client_assertion | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oidc.md]: Same comment here, link to the detailed section.

File: _source/_docs/api/resources/oidc.md [hunk 12]
```diff
@@ -468,26 +495,18 @@ The API takes an Access Token or Refresh Token and revokes it. Revoked tokens ar
...
+| token_type_hint       | A hint of the type of `token`.                                                                                                                                                                                                                        | String |
+| client_id             | The client ID generated as a part of client registration. This is used in conjunction with the `client_secret` parameter to authenticate the client application.                                                                                      | String |
+| client_secret         | The client secret generated as a part of client registration. This is used in conjunction with the `client_id` parameter to authenticate the client application.                                                                                      | String |
+| client_assertion      | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                                                                               | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oidc.md]: Same comment here, link to the detailed section.

File: _source/_docs/api/resources/oidc.md [hunk 13]
```diff
@@ -869,24 +891,26 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
...






[PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.   | String |
+| client_id        | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret`  to authenticate the client application.                                                                                                                                   | String |
+| client_secret    | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.                                                                                                                                    | String |
+| client_assertion | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oidc.md]: Same comment here, link to the detailed section.

File: _source/_docs/api/resources/oauth2.md (hunk 3 above)
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth2.md]: Okay, five nouns in a noun phrase is too many for clarity, so I'll say Client Authentication with Client Secret JWT. Or...? @larsjohansen-okta

File: _source/_docs/api/resources/oauth2.md (hunk 1 above)
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth2.md]: Added to all client_assertion property descriptions in both oidc.md and oauth2.md

File: _source/_docs/api/resources/oauth2.md (hunk 2 above)
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth2.md]: How about this: Use only one of these methods per request or an error will occur.

File: _source/_docs/api/resources/oauth2.md (hunk 5 above)
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth2.md]: I changed it to optional, but isn't it expected that a provided value is validated? Don't we just call out the ones we don't validate if provided?

File: _source/_docs/api/resources/oauth2.md [hunk 14]
```diff
@@ -194,18 +194,20 @@ This API returns Access Tokens, ID Tokens, and Refresh Tokens, depending on the
...







[PKCE](#request-parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request.   | String |
+| client_id             | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret` to authenticate the client application.                                                                                                                                              | String |
+| client_secret         | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.                                                                                                                                              | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "Required if the client has a secret and client credentials are not provided in the Authorization header." is not exactly true anymore as you can provide the proof of knowing the client secret using client_secret_jwt. I'm not sure if I have a great proposal but perhaps you can rewrite it somehow.

File: _source/_docs/api/resources/oauth2.md [hunk 15]
```diff
@@ -194,18 +194,20 @@ This API returns Access Tokens, ID Tokens, and Refresh Tokens, depending on the
...
+| client_assertion      | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`. 

[JWT Details](#token-authentication-methods)                                                                                                                                                                                       | String |
+| client_assertion_type | Indicates a JWT is being used to authenticate a token. Per the 

[Client Authentication spec](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication), the valid value is `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`.                                                                                       | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: "Indicates a JWT is being used to authenticate a token" -> "Indicates a JWT is being used to authenticate a client"

File: _source/_docs/api/resources/oauth2.md [hunk 16]
```diff
@@ -214,17 +216,46 @@ For web and native application types, an additional process is required:
...
+  Authorization: Basic ${Base64(<client_id>:<client_secret>)}
+  ~~~
+* Provide [`client_id`](oidc.html#request-parameters) and [`client_secret`](https://support.okta.com/help/articles/Knowledge_Article/Using-OpenID-Connect)
+  as additional parameters to the POST body (`client_secret_post`)
+* Provide [`client_id`](oidc.html#request-parameters) in a JWT that you sign with the [`client_secret`](https://support.okta.com/help/articles/Knowledge_Article/Using-OpenID-Connect)
+  using HMAC algorithms HS256, HS384, or HS512. Specify the JWT in `client_assertion` and the type, `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`, in `client_assertion_type` in the request. 
+
+Use only one of these methods in a single request or an error will occur.
+
+>If the value returned by `token_endpoint_authentication` in the [OAuth 2.0 Clients API](/docs/api/resources/oauth-clients.html#update-client-application) is not what you wish to use, you can change the value of a client app's `token_endpoint_authentication` with any of the values returned by `token_endpoint_auth_methods_support` (`client_secret_post`, `client_secret_basic`, or `client_secret_jwt`).
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: Replace with ```token_endpoint_authentication ``` with ```token_endpoint_auth_method``` everywhere in this section.

File: _source/_docs/api/resources/oauth2.md (hunk 14 above)
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth2.md]: How about this: "Required if the client has a secret and client credentials are not provided in the Authorization header, and if client_assertion_type is not specified" or "...and if you aren't authenticating the client with a client secret JWT."?

File: _source/_docs/api/resources/oauth2.md [hunk 17]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+Use only one of these methods in a single request or an error will occur.
+
+>If the value returned by `token_endpoint_auth_method` in the [OAuth 2.0 Clients API](/docs/api/resources/oauth-clients.html#update-client-application) is not what you wish to use, you can change the value of a client app's `token_endpoint_auth_method` with any of the values returned by `token_endpoint_auth_methods_support` (`client_secret_post`, `client_secret_basic`, or `client_secret_jwt`).
+
+You can't change this value in the Okta user interface.
+
+##### Request Parameters for Client Authentication with Client Secret JWT
+
+If you use a JWT for token authentication (`client_secret_jwt`), use the following request parameters:
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: request parameters -> token claims

File: _source/_docs/api/resources/oauth2.md [hunk 18]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+| Parameter | Description                                                                      | Type   |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: Parameter -> Claim

File: _source/_docs/api/resources/oauth2.md [hunk 19]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+|:----------|:---------------------------------------------------------------------------------|:-------|
+| exp       | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat       | Optional. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
+| sub       | Required. The subject of the token.                                              | String |
+| aud       | Required. The full URL of the resource you're using the JWT to access.           | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: You're not using the JWT to access a resource, but authenticate a client. 

The full URL of the resource you're using the JWT to access.  -> The full URL of the endpoint you're using the JWT to authenticate to.

File: _source/_docs/api/resources/oauth2.md [hunk 20]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+##### Request Parameters for Client Authentication with Client Secret JWT
+
+If you use a JWT for token authentication (`client_secret_jwt`), use the following request parameters:
+
+| Parameter | Description                                                                      | Type   |
+|:----------|:---------------------------------------------------------------------------------|:-------|
+| exp       | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat       | Optional. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
+| sub       | Required. The subject of the token.                                              | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: We need to mention that the client id needs to be this value

File: _source/_docs/api/resources/oauth2.md [hunk 21]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+| aud       | Required. The full URL of the resource you're using the JWT to access.           | String |
+| iss       | Required. The issuer of the token.                                               | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: We need to mention that the client id needs to be this value

File: _source/_docs/api/resources/oauth2.md [hunk 22]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+##### Token Claims for Client Authentication with Client Secret JWT
+
+If you use a JWT for token authentication (`client_secret_jwt`), use the following token claims:
+
+| Token Claim | Description                                                                      | Type   |
+|:------------|:---------------------------------------------------------------------------------|:-------|
+| exp         | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat         | Optional. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
+| sub         | Required. The subject of the token. This value is the same as the `client_id`.   | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: This value is the same as the `client_id`. -> This value must be the same as the `client_id`.

File: _source/_docs/api/resources/oauth2.md [hunk 23]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+Use only one of these methods in a single request or an error will occur.
+
+>If the value returned by `token_endpoint_auth_method` in the [OAuth 2.0 Clients API](/docs/api/resources/oauth-clients.html#update-client-application) is not what you wish to use, you can change the value of a client app's `token_endpoint_auth_method` with any of the values returned by `token_endpoint_auth_methods_support` (`client_secret_post`, `client_secret_basic`, or `client_secret_jwt`).
+
+You can't change this value in the Okta user interface.
+
+##### Token Claims for Client Authentication with Client Secret JWT
+
+If you use a JWT for token authentication (`client_secret_jwt`), use the following token claims:
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: token authentication -> client authentication

File: _source/_docs/api/resources/oauth2.md [hunk 24]
```diff
@@ -214,17 +216,47 @@ For web and native application types, an additional process is required:
...
+
+| Token Claim | Description                                                                      | Type   |
+|:------------|:---------------------------------------------------------------------------------|:-------|
+| exp         | Required. The expiration time of the token in seconds since January 1, 1970 UTC. | Long   |
+| iat         | Optional. The issuing time of the token in seconds since January 1, 1970 UTC.    | Long   |
+| sub         | Required. The subject of the token. This value is the same as the `client_id`.   | String |
+| aud         | Required. The full URL of the endpoint you're using the JWT to authenticate to.  | String |
+| iss         | Required. The issuer of the token. This value is the same as the `client_id`.    | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: This value is the same as the client_id. -> This value must be the same as the client_id.

File: _source/_docs/api/resources/oauth2.md [hunk 25]
```diff
@@ -353,12 +385,14 @@ An implicit client can only introspect its own tokens, while a confidential clie
...
+| client_assertion      | Required if the `client_assertion_type` is specified. Contains the JWT signed with the `client_secret`. 



[JWT Details](#token-authentication-methods)                                                                                                        | String |
+| client_assertion_type | Indicates a JWT is being used to authenticate a token. Per the 



[Client Authentication spec](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication), the valid value is `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`.        | String |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth2.md]: Indicates a JWT is being used to authenticate a token -> Indicates a JWT is being used to authenticate a client

**Pull request merged**